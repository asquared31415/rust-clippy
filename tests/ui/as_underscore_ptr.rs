//@run-rustfix

#![allow(unused)]
#![deny(clippy::as_underscore_ptr)]

use std::rc::Rc;

// intent: turn a `&u64` into a `*const u8` that points to the bytes of the u64 (⚠️does not work⚠️)
fn owo(x: &u64) -> *const u8 {
    // ⚠️ `&x` is a `&&u64`, so this turns a double pointer into a single pointer
    // ⚠️ This pointer is a dangling pointer to a local
    &x as *const _ as *const u8
}

// note: this test is the same basic idea as above, but uses a `&self` which can be even more
// misleading. In fact this is the case that was found in *real code* (that didn't work)
// that inspired this lint. Make sure that it lints with `&self`!
struct UwU;
impl UwU {
    // like above, this creates a double pointer, and then returns a dangling single pointer
    // intent: turn a `&UwU` into a `*const u8` that points to the same data
    fn as_ptr(&self) -> *const u8 {
        // ⚠️ `&self` is a `&&UwU`, so this turns a double pointer into a single pointer
        // ⚠️ This pointer is a dangling pointer to a local
        &self as *const _ as *const u8
    }
}

fn use_ptr(_: *const ()) {}

fn main() {
    let _: *const u8 = 1 as *const _;
    use_ptr(1 as *const _);

    let _: *mut u8 = 1 as *mut _;

    // Pointer-to-pointer note tests
    // If a _ resolves to a type that is itself a pointer, it's likely a mistake
    // Show a note for all of these cases

    // const ptr to ref
    let r = &&1;
    let _ = r as *const _;

    // const ptr to mut ref
    let r = &&mut 1;
    let _ = r as *const _;

    // mut ptr to ref
    let r = &mut &1;
    let _ = r as *mut _;

    // mut ptr to mut ref
    let r = &mut &mut 1;
    let _ = r as *mut _;

    // ptr to Box
    let b = Box::new(1);
    let r = &b;
    let _ = r as *const _;

    // ptr to Rc
    let rc = Rc::new(1);
    let r = &rc;
    let _ = r as *const _;
}
