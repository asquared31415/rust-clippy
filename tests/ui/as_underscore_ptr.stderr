error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:12:5
   |
LL |     &x as *const _ as *const u8
   |     ^^^^^^--------
   |           |
   |           help: consider giving the type explicitly: `*const &u64`
   |
   = note: the pointed to type is still a pointer-like type
note: the lint level is defined here
  --> $DIR/as_underscore_ptr.rs:4:9
   |
LL | #![deny(clippy::as_underscore_ptr)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:25:9
   |
LL |         &self as *const _ as *const u8
   |         ^^^^^^^^^--------
   |                  |
   |                  help: consider giving the type explicitly: `*const &UwU`
   |
   = note: the pointed to type is still a pointer-like type

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:32:24
   |
LL |     let _: *const u8 = 1 as *const _;
   |                        ^^^^^--------
   |                             |
   |                             help: consider giving the type explicitly: `*const u8`

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:33:13
   |
LL |     use_ptr(1 as *const _);
   |             ^^^^^--------
   |                  |
   |                  help: consider giving the type explicitly: `*const ()`

error: using `as *mut _` conversion
  --> $DIR/as_underscore_ptr.rs:35:22
   |
LL |     let _: *mut u8 = 1 as *mut _;
   |                      ^^^^^------
   |                           |
   |                           help: consider giving the type explicitly: `*mut u8`

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:43:13
   |
LL |     let _ = r as *const _;
   |             ^^^^^--------
   |                  |
   |                  help: consider giving the type explicitly: `*const &i32`
   |
   = note: the pointed to type is still a pointer-like type

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:47:13
   |
LL |     let _ = r as *const _;
   |             ^^^^^--------
   |                  |
   |                  help: consider giving the type explicitly: `*const &mut i32`
   |
   = note: the pointed to type is still a pointer-like type

error: using `as *mut _` conversion
  --> $DIR/as_underscore_ptr.rs:51:13
   |
LL |     let _ = r as *mut _;
   |             ^^^^^------
   |                  |
   |                  help: consider giving the type explicitly: `*mut &i32`
   |
   = note: the pointed to type is still a pointer-like type

error: using `as *mut _` conversion
  --> $DIR/as_underscore_ptr.rs:55:13
   |
LL |     let _ = r as *mut _;
   |             ^^^^^------
   |                  |
   |                  help: consider giving the type explicitly: `*mut &mut i32`
   |
   = note: the pointed to type is still a pointer-like type

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:60:13
   |
LL |     let _ = r as *const _;
   |             ^^^^^--------
   |                  |
   |                  help: consider giving the type explicitly: `*const std::boxed::Box<i32>`
   |
   = note: the pointed to type is still a pointer-like type

error: using `as *const _` conversion
  --> $DIR/as_underscore_ptr.rs:65:13
   |
LL |     let _ = r as *const _;
   |             ^^^^^--------
   |                  |
   |                  help: consider giving the type explicitly: `*const std::rc::Rc<i32>`
   |
   = note: the pointed to type is still a pointer-like type

error: aborting due to 11 previous errors

